### PR TITLE
python312Packages.mlx: upgrade and fix build

### PR DESCRIPTION
**Pull request by Gabriella439**
I polished up the `mlx` package, both upgrading it and also fixing the build

## Things done

- Built on platform(s)
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/op/openmpi/package.nix
Comment (doronbehar) [inline pkgs/by-name/op/openmpi/package.nix:29]: Please separate this to a separate commit.

File: pkgs/by-name/pr/prrte/package.nix
Comment (doronbehar) [inline pkgs/by-name/pr/prrte/package.nix:78]: This too should be a separate commit too.

File: pkgs/development/python-modules/mlx/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/mlx/default.nix:40]: Why not using the standard commit conventions for this upgrade? Please also use the following prefix:

```
python312Packages.mlx
```

And not `python3Packages`, as that doesn't trigger builds by oforg.

File: pkgs/development/python-modules/mlx/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/mlx/default.nix:96]: It'd also be nice to put that in a separate commit.

File: pkgs/by-name/op/openmpi/package.nix
Comment (doronbehar) [inline pkgs/by-name/op/openmpi/package.nix:98]: This should be part of the changes to the `openmpi` changes in the previous commit.

**Review by doronbehar — CHANGES_REQUESTED** [review]
I don't understand the nature of all the changes, but they don't look bad. Only the separation to different commits is very wrong.

**Comment by Gabriella439** [comment]
Also, just to clarify: do you want me to merge these as separate commits or is that just for review purposes?  Typically I use squash merging, which is why I'm asking

File: pkgs/by-name/op/openmpi/package.nix
Comment (markuskowa) [inline pkgs/by-name/op/openmpi/package.nix:29]: What is the motivation to make it configurable? You can of course disable prrte here explicitly, but then it will still be built with the openmpi internally packaged version for prrte (i.e. prrte is never really disabled, only the external dependency). Also: the changes in `mlx` do not make use of this the "disable" feature as far as I can tell.

File: pkgs/development/python-modules/mlx/default.nix
Comment (markuskowa) [reply pkgs/development/python-modules/mlx/default.nix:40]: @doronbehar AFAIK ofborg does not use the commit message to determine the rebuilds, but it calculates the rebuild based on the diff in the nix code (correct me if I am wrong).

File: pkgs/by-name/op/openmpi/package.nix
Comment (Gabriella439) [reply pkgs/by-name/op/openmpi/package.nix:29]: Is there any downside to making it configurable?

Comment (Gabriella439) [reply pkgs/by-name/op/openmpi/package.nix:29]: Also, to clarify: `mlx` does depend on `prrte` being enabled in order to make of use `openmpi`

Without this configure flag enabled, `openmpi` doesn't build and install the `mpirun` executable, which causes the downstream `mlx` build because the MPI configuration step of the `mlx` build looks for the `mpirun` executable.

Comment (doronbehar) [reply pkgs/by-name/op/openmpi/package.nix:29]: This discussion is too messy.

> Is there any downside to making it configurable?

Comment (Gabriella439) [reply pkgs/by-name/op/openmpi/package.nix:29]: I read it, but I don't believe what @markuskowa said is relevant.  You can try setting `enablePRRTE = False` and see for yourself that the OpenMPI support of `mlx` fails, so there is clearly a difference between enabling and disabling it.  If you think I misunderstood something then please elaborate or rephrase instead of restating what was already said (which is not helpful)

File: pkgs/development/python-modules/mlx/default.nix
Comment (doronbehar) [reply pkgs/development/python-modules/mlx/default.nix:40]: > @doronbehar AFAIK ofborg does not use the commit message to determine the rebuilds, but it calculates the rebuild based on the diff in the nix code (correct me if I am wrong).

Nowadays GitHub actions are responsible for detecting the changed packages of a PR by examining the output paths of all packages with and without the PR. However ofborg is also capable of detecting packages that are directly changed by a package, according to the commits' messages' prefixes. See [this](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building). Hopefully, GitHub actions based CI will do that as well, see:

- https://github.com/NixOS/nixpkgs/issues/355847

**Review by doronbehar — CHANGES_REQUESTED** [review]
OK now that the commit order is better (but still not ideal), I understand the purpose of the PR better. As for the Git commit log, I'd do it like this:

```
prrte: enable on Darwin
openmpi: Build with prrte by default, for all platforms
python312Packages.mlx: add abriella439 as maintainer
python312Packages.mlx: 0.18.0 -> 0.21.1; unbreak
```

Note how:

- The packages changed start from the bottom of the dependency tree upwards.
- Each commit prefix starts with a package attribute that is directly handled by Hydra and therefor trigger a build by ofborg thanks to the behavior defined [here](https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building). Specifically, note the difference between `python3Packages` and `python312Packages`. The former is an alias to the later, and Hydra and ofborg really build only the later.
- A package version update like this (to mlx), by nature requires many changes to the expression, and hence there is no point in separating to different commits changes to the mlx expression other then the `meta.maintainers` change. Especially when the package was broken before this PR.

As for the openmpi debate around the prrte change: I'm glad you have discovered that prrte builds fine on Darwin and that openmpi builds fine with it too. I'm also not sure why mlx won't build for you with openmpi compiled with the internal prrte - it'd be interesting to investigate but I don't think it is important enough, because either way, we don't necessarily have to settle this debate, if we'd simply use in `openmpi/package.nix`:

```nix
lib.withFeatureAs true "prrte" (lib.getBin prrte)
```

Also, I expect the `--without-prrte` flag to be unsupported by `openmpi`'s `./configure` script, just like `--without-ofi-libdir` which is mentined there as well.

In general, Nixpkgs favors using our builds of dependencies instead of upstream vendored versions of such dependencies, as it can potentially save disk usage. That's another reason to enable `prrte` unconditionally / use our version of `prrte` unconditionally.

File: pkgs/by-name/op/openmpi/package.nix
Comment (markuskowa) [reply pkgs/by-name/op/openmpi/package.nix:29]: @Gabriella439 Ok, I correct myself: the switch turns off prrte and you are left with a pmix-only version of mpi. 
However, my point here is twofold: The introduction of the "enablePRRTE" feature is unrelated to the topic of this PR (i.e. the improvement of mlx). This is why I asked in the first place.
From your answer, I take it that there is no clear motivation for introducing that switch. If someone comes with a clear argument why a build without PRRTE would be a good idea, I am okay with putting it in.
IMHO we should not add code that has no clear purpose or motivation, as every line of code needs to be maintained in the future (and openmpi is already a big beast).

File: pkgs/by-name/pr/prrte/package.nix
Comment (doronbehar) [inline pkgs/by-name/pr/prrte/package.nix:78]: I once heard someone saying that `platforms.unix` includes BSD, which is not necessarily a supported platform :). However, for most practical purposes, this is the same, as Hydra doesn't build for BSD ;). Just noting it here for you so you'd know for next time. Don't worry this is not a blocker for the PR.

**Review by doronbehar — APPROVED** [review]
Changes look great, including the commit messages are very well undestood. Thanks for the contribution! I'll give @markuskowa a day or two to approve / merge.

**Review by markuskowa — APPROVED** [review]
Thanks! LGTM.

Side note: When I package new packages (such as PRRTE), I only enable them for Linux usually. Darwin support is difficult to handle if you do not have access to a Darwin machine. I thus leave it to Darwin user/devs to enable Darwin support on packages.

**Pull request merged**
